hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSnapshotBackgroundServices.java

```diff
@@ -0,0 +1,722 @@
+    if (cluster != null) {
+      cluster.shutdown();
+    }
+  }
+
+  @Test
+  @DisplayName("testSnapshotAndKeyDeletionBackgroundServices")
+  @SuppressWarnings("methodlength")
+  public void testSnapshotAndKeyDeletionBackgroundServices()
+      throws IOException, InterruptedException, TimeoutException {
+    OzoneManager leaderOM = getLeaderOM();
+    OzoneManager followerOM = getInactiveFollowerOM(leaderOM);
+
+    createSnapshotsEachWithNewKeys(leaderOM);
+
+    startInactiveFollower(leaderOM, followerOM, () -> {
+    });
+
+    // Read & Write after snapshot installed.
+    List<String> newKeys = writeKeys(1);
+    readKeys(newKeys);
+
+    OzoneManager newLeaderOM =
+        getNewLeader(leaderOM, followerOM.getOMNodeId(), followerOM);
+    OzoneManager newFollowerOM =
+        cluster.getOzoneManager(leaderOM.getOMNodeId());
+    Assertions.assertEquals(leaderOM, newFollowerOM);
+
+    SnapshotInfo newSnapshot = createOzoneSnapshot(newLeaderOM,
+        SNAPSHOT_NAME_PREFIX + RandomStringUtils.randomNumeric(5));
+
+      /*
+      Check whether newly created key data is reclaimed
+      create key a
+      create snapshot b
+      delete key a
+      create snapshot c
+      assert that a is in c's deleted table
+      create snapshot d
+      delete snapshot c
+      wait until key a appears in deleted table of d.
+    */
+    // create key a
+    String keyNameA = writeKeys(1).get(0);
+    String keyA = OM_KEY_PREFIX + ozoneBucket.getVolumeName() +
+        OM_KEY_PREFIX + ozoneBucket.getName() +
+        OM_KEY_PREFIX + keyNameA;
+    Table<String, OmKeyInfo> omKeyInfoTable = newLeaderOM
+        .getMetadataManager()
+        .getKeyTable(ozoneBucket.getBucketLayout());
+    OmKeyInfo keyInfoA = omKeyInfoTable.get(keyA);
+    Assertions.assertNotNull(keyInfoA);
+
+    // create snapshot b
+    SnapshotInfo snapshotInfoB = createOzoneSnapshot(newLeaderOM,
+        SNAPSHOT_NAME_PREFIX + RandomStringUtils.randomNumeric(5));
+    Assertions.assertNotNull(snapshotInfoB);
+
+    // delete key a
+    ozoneBucket.deleteKey(keyNameA);
+
+    GenericTestUtils.waitFor(() -> {
+      try {
+        return Objects.isNull(omKeyInfoTable.get(keyA));
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }, 1000, 10000);
+
+    // create snapshot c
+    SnapshotInfo snapshotInfoC = createOzoneSnapshot(newLeaderOM,
+        SNAPSHOT_NAME_PREFIX + RandomStringUtils.randomNumeric(5));
+
+    // get snapshot c
+    OmSnapshot snapC;
+    try (ReferenceCounted<IOmMetadataReader, SnapshotCache> rcC = newLeaderOM
+        .getOmSnapshotManager()
+        .checkForSnapshot(volumeName, bucketName,
+            getSnapshotPrefix(snapshotInfoC.getName()), true)) {
+      Assertions.assertNotNull(rcC);
+      snapC = (OmSnapshot) rcC.get();
+    }
+
+    // assert that key a is in snapshot c's deleted table
+    GenericTestUtils.waitFor(() -> {
+      try (TableIterator<String, ? extends Table.KeyValue<String,
```

Comment:
1. What's the reason to iterate over the table? Can we use `get(keyA)`, `getSkipCache(keyA)` or `getIfExist(keyA)`?  Something like below?

```
    GenericTestUtils.waitFor(() -> {
      try {
        return snapC.getMetadataManager().getDeletedTable().getIfExist(keyA)
            != null;
      } catch (IOException e) {
        Assertions.fail();
        return false;
      }
    }, 1000, 10000);
```

2. This repetitive code and similar code is written at 329-343. May be something like if you don't care about the value type..
```
  GenericTestUtils.waitFor(() ->
        checkForKeyInTable(snapC.getMetadataManager().getDeletedTable(), keyA),
    1000, 10000);  
  ...
  ...
  private <V> boolean checkForKeyInTable(Table<String, V> table,
                                         String key) {
    try (TableIterator<String, ? extends Table.KeyValue<String, V>> iterator
             = table.iterator()) {
      while (iterator.hasNext()) {
        Table.KeyValue<String, V> next = iterator.next();
        if (next.getKey().contains(key)) {
          return true;
        }
      }
      return false;
    } catch (IOException e) {
      return false;
    }
  }
```

Reply:
I tried using mentioned methods. However, all of those return null. Because of that I've decided to iterate over table. I refactored it to be more readable and reusable according to your suggestion.